### PR TITLE
chore: wl fiat

### PR DESCRIPTION
**Pull request by catngh**
**What does this PR do?**

- New service from nghenhan API
- Add field `is_fiat` to `user_watchlist_items`

### CODE REVIEW

File: pkg/entities/defi.go [hunk 1]
```diff
@@ -622,22 +623,44 @@ func (e *Entity) GetUserWatchlist(req request.GetUserWatchlistRequest) (*[]respo
...
-			latestPrice := item.SparkLineIn7d.Price[len(item.SparkLineIn7d.Price)-1]
-			oldPrice := item.SparkLineIn7d.Price[0]
-			item.CurrentPrice = latestPrice
-			item.PriceChangePercentage7dInCurrency = (latestPrice - oldPrice) / oldPrice * 100
+		item := response.CoinMarketItemData{}
+		// process fiat data
+		if pair.IsFiat {
+			fiatData, err := e.svc.Nghenhan.GetFiatHistoricalChart(queries[0], queries[1], "w", 10)
+			if err != nil {
+				e.log.Fields(logger.Fields{"pair": pair}).Error(err, "[entity.GetUserWatchlist] entity.GetFiatHistoricalExchangeRates() failed")
```
Comment (hoangharry) [inline pkg/entities/defi.go:631]: not the right function in log

**Review by hoangharry — APPROVED** [review]

File: pkg/entities/defi.go [hunk 2]
```diff
@@ -622,22 +623,44 @@ func (e *Entity) GetUserWatchlist(req request.GetUserWatchlistRequest) (*[]respo
...
+				e.log.Fields(logger.Fields{"pair": pair}).Error(err, "[entity.GetUserWatchlist] Nghenhan.GetFiatHistoricalChart failed")
+				return nil, err
+			}
+			fiatRate := []float64{}
+			for _, v := range fiatData.Data {
+				cPrice, _ := strconv.ParseFloat(v.ClosePrice, 64)
+				fiatRate = append(fiatRate, cPrice)
+			}
+
+			lastestPrice := fiatRate[len(fiatRate)-1]
```
Comment (nguyennh4522) [inline pkg/entities/defi.go:640]: should check len(fiatRate) > 0 before getting last value.

File: pkg/entities/defi.go [hunk 3]
```diff
@@ -622,22 +623,44 @@ func (e *Entity) GetUserWatchlist(req request.GetUserWatchlistRequest) (*[]respo
...
+			oldPrice := fiatRate[0]
```
Comment (nguyennh4522) [inline pkg/entities/defi.go:641]: same as above

File: pkg/entities/defi.go [hunk 4]
```diff
@@ -622,22 +623,44 @@ func (e *Entity) GetUserWatchlist(req request.GetUserWatchlistRequest) (*[]respo
...
+			item.PriceChangePercentage7dInCurrency = (lastestPrice - oldPrice) / oldPrice * 100
```
Comment (nguyennh4522) [inline pkg/entities/defi.go:642]: app will crash if `oldPrice = 0`

File: pkg/request/defi.go [hunk 5]
```diff
@@ -42,6 +42,7 @@ type AddToWatchlistRequest struct {
 	UserID      string `json:"user_id"`
 	Symbol      string `json:"symbol"`
 	CoinGeckoID string `json:"coin_gecko_id"`
+	IsFiat      bool   `json:"is_fiat" form:"is_fiat,default=false"`
```
Comment (nguyennh4522) [inline pkg/request/defi.go:45]: default value of `bool` always is false

**Review by nguyennh4522 — APPROVED** [review]

**Review by nguyennh4522 — APPROVED** [review]

**Pull request merged**